### ISSUE TITLE
Reduce the minimum buffer size required by the StreamWriter class

### ISSUE THREAD

**Issue by GSPP** [dotnet/corefx#8376]
`StreamWriter` forces its `bufferSize` constructor parameter to be at least `128`. There is no workaround. This seems unnecessarily limiting. `bufferSize = 128` actually consumes about `256+(128*3)` bytes of memory with UTF8 (I checked this using the debugger). This is a significant amount of memory per instance.

If the user really wants a smaller size, maybe knowing something about the data distribution, the user-specified value should take effect. For example, I might know that I will be writing only a few dozen bytes when serializing a cookie or an HTTP header.

Simply lowering this minimum value would be a compatibility concern. So I propose this:

The user-specified `bufferSize` value should be used with a minimum of `1`. Each time the buffer is actually exhausted, grow the internal buffers by doubling them each time until a size of `128` is reached.

This way the user can specify a smaller buffer size if they know that not much data will be coming. But if a small size was mistakenly specified `StreamWriter` will quickly adapt to a bigger, more efficient buffer. This mitigates the compatibility concern.

An alternative scheme would be to always start with a size of `16` and double until `max(bufferSize, 128)` is reached. That way the user-provided value is taken as a maximum instead of a precise value. I'm not convinced this scheme should be implemented. Just mentioning it.

**Comment by svick** [comment]
> Simply lowering this minimum value would be a compatibility concern.

Why? I can see it mattering for .Net Framework, where you want, say, `bufferSize: 1` used incorrectly to keep being reasonably fast. But for .Net Core, I don't think the compatibility bar should be that high, especially since I don't think this is going to be a common issue.

The increasing buffer solution might work, but I don't think it would be right to double after every exhaustion: I think that allocates too much.

**Comment by svick** [comment]
Also, since what I think you actually want is to avoid allocations, maybe some way to make `StreamWriter` work with `System.Buffers` could help? (But [it can't be used directly](https://github.com/dotnet/corefx/issues/5598#issuecomment-173726925).)

**Comment by GSPP** [comment]
>  I can see it mattering for .Net Framework, where you want, say, bufferSize: 1 used incorrectly to keep being reasonably fast.

That's the concern. Maybe we could increase to 128 immediately on the first exhaustion. This would cause only one "false" allocation in the bad case.

I think the exponentially increasing buffer size in quite a useful behavior, even if the user does not specify any size. It's like `List<T>` and it's an amazingly effective heuristic in practice.

> But for .Net Core ...

While that is true I believe that the only long-term feasible approach is to keep the .NET flavors closely aligned in semantics *and* in code. Divergence causes user confusion and makes it gradually harder to cross-apply patches.

I don't know what the official stance is on that but since I also don't know how *else* to proceed I assume that is the plan.

**Comment by ianhays** [comment]
Wouldn't that be somewhat odd to increase the size like that though? We'd still be disallowing sizes lower than 128, we'd just be be doing it in a more hidden way (waiting for a write that won't fit the small buffer).

I don't know that the added complexity and the slowdown of the ```bufferSize=1``` case are justified by allowing buffers lower than 128. If not for the compat concerns (i.e. if we were creating this API instead of modifying it), I think it would make the most sense to allow *any* buffer size regardless of the possible inefficiency. 

>I believe that the only long-term feasible approach is to keep the .NET flavors closely aligned in semantics and in code. Divergence causes user confusion and makes it gradually harder to cross-apply patches.

>I don't know what the official stance is on that but since I also don't know how else to proceed I assume that is the plan.

Good call! That's generally the plan, yes. We're very careful about accepting breaking changes (i.e. changes that will cause code to break when porting from the full framework to .net core).

So, summing up here. 

Pros:
- More developer control. Can set whatever buffer size you want (>1).
- Saves memory when a dev only needs a very small buffer


Cons:
- Slowdown for ported code where buffersize==1 while it spins up the buffer size to 128
- Adds new complexity
- We're still enforcing a (undocumented) minimum buffer, we just wait until later to enact that rule

Thoughts?

**Comment by benaadams** [comment]
Overload to pass a buffer in?

**Comment by GSPP** [comment]
Maybe an externally provided buffer is the way to go (as suggested by benaadams). Problem is there are two internal buffers so that does not work right away. But that would be really nice to conserve memory even across writers. I could see that idea being used for FileStream, BufferedStream and StreamReader as well!

**Comment by benaadams** [comment]
@GSPP is basically what aspnetcore are doing https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.WebUtilities/HttpResponseStreamWriter.cs and https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.WebUtilities/HttpRequestStreamReader.cs

**Comment by ianhays** [comment]
Nice thing about a ArrayPool overload (like in the files linked) is that is would be useful in other scenarios, not just when the bufferSize < 128.

The problem with that would be that StreamWriter is in mscorlib which can not take a dependency on System.Buffers, so if anything it would have to take two ref arrays.

**Comment by GSPP** [comment]
Why does System.Buffers have to be separate? Is it so big? Maybe just move it into mscorlib. There's lots of IO stuff in mscorlib that wants to take advantage.

**Comment by ianhays** [comment]
>Why does System.Buffers have to be separate? Is it so big? Maybe just move it into mscorlib. There's lots of IO stuff in mscorlib that wants to take advantage.

I vaguely recall some discussion around that when @stephentoub was applying System.Buffers to existing libraries. It's not so much a matter of size as it is a matter of design principle. Mscorlib contains only the 'core' libraries, particularly those used in the runtime. IMO System.Buffers, being platform-agnostic, makes more sense as a standalone assembly in CoreFX - though I agree with you that it's unfortunate it cannot be used in API as a result of that.

**Comment by GSPP** [comment]
One could argue that IO should not be in mscorlib. It could be reduced to be the "interface to the runtime". Not sure if that's achievable...

**Comment by ianhays** [comment]
In https://github.com/dotnet/corefx/commit/8d73384985d5a261e7a721acc13f5557ab67cba5 I moved StreamReader and StreamWriter out of mscorlib. This unblocks the ArrayPool overload we were talking about earlier.

Since the discussion has diverged from the OP and we're now talking about adding a new API instead of modifying the existing one, I'm going to close out this issue. The next step would be to open a new issue with a formal API proposal for a StreamWriter/Reader with ArrayPool overloads.
